src/Interpreters/Cache/WriteBufferToFileSegment.cpp

```diff
@@ -0,0 +1,94 @@
+#include <Interpreters/Cache/WriteBufferToFileSegment.h>
+
+#include <Common/logger_useful.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int NOT_ENOUGH_SPACE;
+    extern const int LOGICAL_ERROR;
+}
+
+namespace
+{
+    class SwapHelper
+    {
+    public:
+        SwapHelper(WriteBuffer & b1_, WriteBuffer & b2_) : b1(b1_), b2(b2_) { b1.swap(b2); }
+        ~SwapHelper() { b1.swap(b2); }
+
+    private:
+        WriteBuffer & b1;
+        WriteBuffer & b2;
+    };
+}
+
+WriteBufferToFileSegment::WriteBufferToFileSegment(std::unique_ptr<WriteBuffer> impl_, FileSegment * file_segment_)
+    : WriteBufferFromFileDecorator(std::move(impl_)), file_segment(file_segment_)
+{
+    auto downloader = file_segment->getOrSetDownloader();
+    if (downloader != FileSegment::getCallerId())
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Failed to set a downloader. ({})", file_segment->getInfoForLog());
+}
+
+void WriteBufferToFileSegment::nextImpl()
+{
+    size_t bytes_to_write = offset();
+    LOG_WARNING(&Poco::Logger::get("WriteBufferToFileSegment"), "Writing {} bytes", bytes_to_write);
+
+    bool ok = file_segment->reserve(bytes_to_write);
+    if (!ok)
+        throw Exception(ErrorCodes::NOT_ENOUGH_SPACE, "Failed to reserve space for the file cache ({})", file_segment->getInfoForLog());
+
+    try
+    {
```

Comment:
but if there was an exception from `reserve` we also need to reset downloader same as in catch below. But also probably we do not need to do that in both cases as downloader should be reset in `complete()` method when `holder` calls it from its destructor and sees that this holder is the last owner of the segment. So I'd better just to be consistent either reset downloader manually in both cases, or not reset at all.

Same with `completeWithState`.

Reply:
Removed this calls and added comments